libtenzir/builtins/operators/from_to_2.cpp

```diff
@@ -1107,62 +1121,61 @@ class from_file_impl {
       });
   }
 
-  from_file_actor::pointer self_;
+  auto register_metrics(uint64_t nested_operator_index, uuid nested_metrics_id,
+                        type schema) -> caf::result<void> {
+    auto& id = registered_metrics_[nested_operator_index][nested_metrics_id];
+    id = uuid::random();
+    return self_->mail(operator_index_, id, std::move(schema))
+      .delegate(metrics_receiver_);
+  }
 
-  // TODO
+  auto handle_metrics(uint64_t nested_operator_index, uuid nested_metrics_id,
+                      record metrics) -> caf::result<void> {
+    const auto& id
+      = registered_metrics_[nested_operator_index][nested_metrics_id];
+    return self_->mail(operator_index_, id, std::move(metrics))
+      .delegate(metrics_receiver_);
+  }
+
+  from_file_actor::pointer self_;
   std::unique_ptr<diagnostic_handler> dh_;
+  std::shared_ptr<arrow::fs::FileSystem> fs_;
+
+  // The configuration and things derived from it.
+  from_file_args args_;
   event_order order_;
+  glob glob_;
+  std::string root_path_;
 
-  // watching
+  // Watching is implemented by checking against the files seen previously.
   file_set previous_;
   file_set current_;
 
-  // communication with bridges
+  // Communication with the operator bridges.
   std::deque<caf::typed_response_promise<table_slice>> gets_;
   std::deque<std::pair<table_slice, caf::typed_response_promise<void>>> puts_;
 
-  // configuration
-  located<std::string> url_;
-  bool watch_;
-  bool remove_;
-  std::optional<ast::field_path> path_field_;
-  std::optional<located<pipeline>> pipe_;
-
-  // things derived from configuration
-  glob glob_;
-  std::string root_path_;
-
-  // stuff for spawning subpipelines
+  // Information needed for spawning subpipelines.
   std::string definition_;
   node_actor node_;
   bool is_hidden_;
 
-  // job management
-  size_t parallel_ = 10;
+  // Job management.
   size_t active_jobs_ = 0;
   std::deque<arrow::fs::FileInfo> jobs_;
-  bool spawned_all_jobs_ = false;
+  bool added_all_jobs_ = false;
 
-  // arrow stuff
-#if USE_EXECUTOR
-  caf_executor io_executor_;
-  arrow::io::IOContext io_ctx_;
-#endif
-  std::shared_ptr<arrow::fs::FileSystem> fs_;
+  // Forwarding metrics.
+  uint64_t operator_index_ = 0;
+  detail::flat_map<uint64_t, detail::flat_map<uuid, uuid>> registered_metrics_;
```

Comment:
This shouldn't be needed. The UUIDs are already random, so you can just forward them as-is.

Reply:
Hmm, I just lazily copied this from `if`. Why would it be different between the two?

Reply:
I think in `if` I too lazily adapted this when changing the metrics id to no longer be an integer that is counted up but instead a random UUID. Might wanna make that change in there as well.

Reply:
Done